compiler/src/dotty/tools/dotc/transform/sjs/ExplicitJSClasses.scala

```diff
@@ -12,54 +17,728 @@ import core.Types._
+ *    desugars into
+ *    `withContextualJSClassValue(js.constructorOf[C], D.super[C].m[...Ts](...args))`.
+ *
+ *  For any nested JS *object*, their (only) instantiation point of the form
+ *  `new O$()` is rewritten as
+ *  `withContextualJSClassValue(js.constructorOf[ParentClassOfO], new O$())`,
+ *  so that the back-end receives a reified reference to the parent class of
+ *  `O`.
+ *
+ *  TODO A similar treatment is applied on anonymous JS classes, which
+ *  basically define something very similar to an `object`, although without
+ *  its own JS class.
+ *
+ *  --------------------------------------
+ *
+ *  (E) Desugar `x.isInstanceOf[C]` for nested JS classes
  *
- *  - Rewrite `js.constructorOf[T]` into `scala.scalajs.runtime.constructorOf(classOf[T])`,
- *    where the `classOf[T]` is represented as a `Literal`.
+ *  They are desugared into `js.special.instanceof(x, js.constructorOf[C])`.
+ *
+ *  --------------------------------------
+ *
+ *  (F) Desugar `js.constructorOf[C]`
+ *
+ *  Finally, this phase rewrites all calls to `js.constructorOf[C]`, including
+ *  the ones generated by the previous steps. The transformation depends on the
+ *  nature of `C`:
+ *
+ *  - If `C` is a statically accessible class, desugar to
+ *    `runtime.constructorOf(classOf[C])` so that the reified symbol survives
+ *    erasure and reaches the back-end.
+ *  - If `C` is an inner JS class, it must be of the form `path.D` for some
+ *    pair (`path`, `D`), and we desugar it to `path.D\$jsclass`, using the
+ *    field created by step (A) (it is an error if `C` is of the form
+ *    `Enclosing#D`).
+ *  - If `C` is a local JS class, desugar to `C\$jsclass`, using the local val
+ *    created by step (C).
  */
 class ExplicitJSClasses extends MiniPhase with InfoTransformer { thisPhase =>
   import ExplicitJSClasses._
   import ast.tpd._
 
   override def phaseName: String = ExplicitJSClasses.name
 
+  private var MyState: Store.Location[MyState] = _
+  private def myState(using Context) = ctx.store(MyState)
+
+  override def initContext(ctx: FreshContext): Unit =
+    MyState = ctx.addLocation[MyState]()
+
   override def isEnabled(using Context): Boolean =
     ctx.settings.scalajs.value
 
   override def runsAfter: Set[String] = Set(PatternMatcher.name, HoistSuperArgs.name)
 
   override def changesMembers: Boolean = true // the phase adds fields for inner JS classes
 
-  override def transformInfo(tp: Type, sym: Symbol)(using Context): Type = {
-    // Currently we don't do anything here. Eventually we'll add fields for inner JS classes.
-    tp
+  /** Is the given symbol an owner for which this transformation applies?
+   *
+   *  This applies if either or both of the following are true:
+   *  - It is not a static owner, or
+   *  - It is a non-native JS object.
+   *
+   *  The latter is necessary for scala-js/scala-js#4086.
+   */
+  private def isApplicableOwner(sym: Symbol)(using Context): Boolean = {
```

Comment:
Self-review: change the name. It only tells whether the owner is applicable for `transformInfo`/`transformTemplate`, i.e., whether it should receive `$jsclass` and/or `$jsobject` fields.